// @yfme/weapp-dev-mcp/dist/fastmcp.js
import { FastMCP, FastMCPSession, UnexpectedStateError, UserError, audioContent, imageContent, } from "fastmcp";
const sessionProto = FastMCPSession.prototype;
if (!sessionProto.__weappDevPatchedCompleteHandlers) {
    const originalSetupCompleteHandlers = sessionProto.setupCompleteHandlers;
    sessionProto.setupCompleteHandlers = function patchedSetupCompleteHandlers() {
        try {
            originalSetupCompleteHandlers.call(this);
        }
        catch (error) {
            if (error instanceof Error &&
                error.message.includes("Server does not support completions")) {
                console.warn("[weapp-dev-mcp] FastMCP completion handlers were not registered because the server " +
                    "did not advertise completion capabilities. This server only exposes tools, so " +
                    "completions are optional and the error is safely ignored.");
                return;
            }
            throw error;
        }
    };
    sessionProto.__weappDevPatchedCompleteHandlers = true;
}
export { FastMCP, UnexpectedStateError, UserError, audioContent, imageContent, };
