// @yfme/weapp-dev-mcp/dist/weappClient.js
import { UserError } from "fastmcp";
import automator from "miniprogram-automator";
import { ConfigError, resolveConfig, } from "./config.js";
export class WeappAutomatorManager {
    miniProgram;
    config;
    consoleLogs = [];
    maxLogs = 1000; // 最多保存1000条日志
    getConsoleLogs() {
        return [...this.consoleLogs];
    }
    clearConsoleLogs() {
        this.consoleLogs = [];
    }
    async withMiniProgram(log, options, handler) {
        const { overrides, reconnect } = options;
        let config;
        try {
            config = resolveConfig(overrides, this.config);
        }
        catch (error) {
            if (error instanceof ConfigError) {
                throw new UserError(error.message);
            }
            throw error;
        }
        if (reconnect) {
            await this.close(log);
        }
        const canReuse = this.miniProgram && this.config && isSameConfig(this.config, config);
        if (!canReuse) {
            await this.close(log);
            log.info("Establishing WeChat DevTools automation session", {
                mode: config.mode,
                projectPath: config.projectPath,
                wsEndpoint: config.wsEndpoint,
                port: config.port,
            });
            try {
                this.miniProgram = await this.connect(config);
                this.config = config;
                this.attachLogging(this.miniProgram, log);
            }
            catch (error) {
                this.miniProgram = undefined;
                this.config = undefined;
                const message = error instanceof Error ? error.message : String(error);
                throw new UserError(`Failed to ${config.mode === "connect" ? "connect to" : "launch"} WeChat DevTools: ${message}`);
            }
        }
        const activeProgram = this.miniProgram;
        try {
            return await handler(activeProgram, config);
        }
        finally {
            if (config.autoClose) {
                await this.close(log);
            }
        }
    }
    async withPage(log, options, handler) {
        return this.withMiniProgram(log, options, async (miniProgram, config) => {
            const page = await miniProgram.currentPage();
            if (!page) {
                throw new UserError("Mini Program page stack is empty. Ensure the project window is open.");
            }
            return handler(page, miniProgram, config);
        });
    }
    async close(log) {
        if (!this.miniProgram) {
            return;
        }
        try {
            if (this.config?.mode === "launch") {
                await this.miniProgram.close();
            }
            else {
                this.miniProgram.disconnect();
            }
            log?.debug("Closed WeChat DevTools automation session");
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            log?.warn("Failed to close WeChat DevTools cleanly", { message });
        }
        finally {
            this.miniProgram.removeAllListeners();
            this.miniProgram = undefined;
            this.config = undefined;
        }
    }
    async connect(config) {
        if (config.mode === "connect") {
            return automator.connect({ wsEndpoint: config.wsEndpoint });
        }
        return automator.launch({
            cliPath: config.cliPath,
            projectPath: config.projectPath,
            timeout: config.timeout,
            port: config.port,
            account: config.account,
            ticket: config.ticket,
            trustProject: config.trustProject,
            args: config.args,
            cwd: config.cwd,
        });
    }
    attachLogging(miniProgram, log) {
        miniProgram.on("console", (event) => {
            const serialized = toSerializable(event);
            const args = event?.args;
            const logEntry = {
                type: typeof event?.type === "string" ? event.type : "log",
                message: Array.isArray(args) ? args.map(arg => typeof arg === "string" ? arg : JSON.stringify(arg)).join(" ") : String(serialized),
                timestamp: Date.now(),
                data: serialized,
            };
            // 保存日志，限制数量
            this.consoleLogs.push(logEntry);
            if (this.consoleLogs.length > this.maxLogs) {
                this.consoleLogs.shift();
            }
            log.debug("Mini Program console event", {
                event: serialized,
            });
        });
        miniProgram.on("exception", (event) => {
            const serialized = toSerializable(event);
            const logEntry = {
                type: "exception",
                message: typeof event?.message === "string" ? event.message : String(serialized),
                timestamp: Date.now(),
                data: serialized,
            };
            // 保存异常日志
            this.consoleLogs.push(logEntry);
            if (this.consoleLogs.length > this.maxLogs) {
                this.consoleLogs.shift();
            }
            log.error("Mini Program exception", {
                event: serialized,
            });
        });
    }
}
function toSerializable(value) {
    if (value === null || value === undefined) {
        return value;
    }
    if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
        return value;
    }
    if (Array.isArray(value)) {
        return value.map((item) => toSerializable(item));
    }
    if (typeof value === "object") {
        const entries = Object.entries(value).map(([key, val]) => [key, toSerializable(val)]);
        return Object.fromEntries(entries);
    }
    return String(value);
}
function isSameConfig(a, b) {
    return (a.mode === b.mode &&
        a.cliPath === b.cliPath &&
        a.projectPath === b.projectPath &&
        a.wsEndpoint === b.wsEndpoint &&
        a.timeout === b.timeout &&
        a.port === b.port &&
        a.account === b.account &&
        a.ticket === b.ticket &&
        a.trustProject === b.trustProject &&
        a.cwd === b.cwd &&
        a.autoClose === b.autoClose &&
        areArgsEqual(a.args, b.args));
}
function areArgsEqual(a, b) {
    if (!a && !b) {
        return true;
    }
    if (!a || !b) {
        return false;
    }
    if (a.length !== b.length) {
        return false;
    }
    return a.every((value, index) => value === b[index]);
}
