// @yfme/weapp-dev-mcp/dist/tools/common.js
import { UserError, } from "fastmcp";
import { z } from "zod";
import { connectionOverridesSchema } from "../config.js";
export const connectionContainerSchema = z.object({
    connection: connectionOverridesSchema.optional(),
});
export const connectionOnlyParameters = connectionContainerSchema;
export const ensureConnectionParameters = connectionContainerSchema
    .extend({
    reconnect: z.coerce.boolean().optional().default(false),
});
export const querySchema = z.record(z.string()).optional();
export const stringListSchema = z
    .union([z.string(), z.array(z.string()), z.undefined()])
    .transform((value) => {
    if (!value) {
        return undefined;
    }
    const list = Array.isArray(value) ? value : value.split(/\s+/);
    const normalized = list.map((item) => item.trim()).filter(Boolean);
    return normalized.length ? normalized : undefined;
});
export function buildUrl(path, query) {
    const normalizedPath = path.startsWith("/") ? path : `/${path}`;
    if (!query || Object.keys(query).length === 0) {
        return normalizedPath;
    }
    const searchParams = new URLSearchParams(query);
    const separator = normalizedPath.includes("?") ? "&" : "?";
    const search = searchParams.toString();
    return search ? `${normalizedPath}${separator}${search}` : normalizedPath;
}
export function formatJson(value) {
    const serialized = JSON.stringify(value, null, 2);
    return serialized ?? String(value);
}
export function toTextResult(text) {
    return {
        content: [
            {
                type: "text",
                text,
            },
        ],
    };
}
export async function readNamedValues(names, reader, kind) {
    if (!names?.length) {
        return undefined;
    }
    const entries = [];
    for (const name of names) {
        try {
            entries.push([name, await reader(name)]);
        }
        catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            entries.push([name, `Failed to read ${kind}: ${message}`]);
        }
    }
    return Object.fromEntries(entries);
}
export async function resolveElement(page, selector, innerSelector) {
    if (!page || typeof page.$ !== "function") {
        throw new UserError("Page instance is not available to resolve elements.");
    }
    let element = await page.$(selector);
    if (!element) {
        throw new UserError(`Element not found for selector "${selector}".`);
    }
    if (innerSelector) {
        if (typeof element.$ !== "function") {
            throw new UserError(`Element for selector "${selector}" does not support nested queries.`);
        }
        const inner = await element.$(innerSelector);
        if (!inner) {
            throw new UserError(`Element not found for selector "${innerSelector}" within "${selector}".`);
        }
        element = inner;
    }
    return element;
}
export async function summarizeElement(element, options) {
    const tagName = typeof element?.tagName === "string" ? element.tagName : null;
    const withWxml = options?.withWxml ?? false;
    const [text, value, outerWxml, size, offset, scrollWidth, scrollHeight] = await Promise.all([
        typeof element?.text === "function"
            ? element.text().catch(() => null)
            : null,
        typeof element?.value === "function"
            ? element.value().catch(() => null)
            : null,
        withWxml && typeof element?.outerWxml === "function"
            ? element.outerWxml().catch(() => null)
            : null,
        typeof element?.size === "function"
            ? element.size().catch(() => null)
            : null,
        typeof element?.offset === "function"
            ? element.offset().catch(() => null)
            : null,
        typeof element?.scrollWidth === "function"
            ? element.scrollWidth().catch(() => null)
            : null,
        typeof element?.scrollHeight === "function"
            ? element.scrollHeight().catch(() => null)
            : null,
    ]);
    const result = {
        tagName: toSerializableValue(tagName),
        text: toSerializableValue(text),
        value: toSerializableValue(value),
        size: toSerializableValue(size),
        offset: toSerializableValue(offset),
    };
    // 当 withWxml 为 true 时，返回完整的 outerWxml
    if (withWxml && outerWxml !== null) {
        result.outerWxml = toSerializableValue(outerWxml);
    }
    // scroll-view 专用属性，仅在有值时添加
    if (scrollWidth !== null) {
        result.scrollWidth = toSerializableValue(scrollWidth);
    }
    if (scrollHeight !== null) {
        result.scrollHeight = toSerializableValue(scrollHeight);
    }
    return result;
}
export async function waitOnPage(page, waitMs) {
    if (!waitMs) {
        return;
    }
    if (page && typeof page.waitFor === "function") {
        await page.waitFor(waitMs);
    }
}
export function serializePageSummary(page) {
    if (!isPageLike(page)) {
        return toSerializableValue(page);
    }
    const summary = {
        path: page.path,
    };
    if (page.query !== undefined) {
        summary.query = toSerializableValue(page.query);
    }
    return summary;
}
export function toSerializableValue(value) {
    if (value === null || value === undefined) {
        return value;
    }
    if (typeof value === "string" ||
        typeof value === "number" ||
        typeof value === "boolean") {
        return value;
    }
    if (typeof value === "bigint") {
        return value.toString();
    }
    if (value instanceof Date) {
        return value.toISOString();
    }
    if (typeof Buffer !== "undefined" && Buffer.isBuffer(value)) {
        return value.toString("base64");
    }
    if (Array.isArray(value)) {
        return value.map((item) => toSerializableValue(item));
    }
    if (isPageLike(value)) {
        return serializePageSummary(value);
    }
    if (typeof value === "object") {
        const entries = Object.entries(value).map(([key, val]) => [key, toSerializableValue(val)]);
        return Object.fromEntries(entries);
    }
    return String(value);
}
function isPageLike(value) {
    return (value !== null &&
        typeof value === "object" &&
        typeof value.path === "string");
}
export function createFunctionFromSource(source, context) {
    try {
        const fn = new Function(`return (${source});`)();
        if (typeof fn !== "function") {
            throw new Error("Source did not evaluate to a function.");
        }
        return fn;
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        throw new UserError(`${context} is invalid: ${message}`);
    }
}
